{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script>jQuery(function() {if (jQuery(\"body.notebook_app\").length == 0) { jQuery(\".input_area\").toggle(); jQuery(\".prompt\").toggle();}});</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<button onclick=\"jQuery('.input_area').toggle(); jQuery('.prompt').toggle();\">Toggle code</button>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\zhila.esnaashari\\Anaconda3\\lib\\site-packages\\sklearn\\preprocessing\\data.py:645: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.\n",
      "  return self.partial_fit(X, y)\n",
      "C:\\Users\\zhila.esnaashari\\Anaconda3\\lib\\site-packages\\sklearn\\base.py:464: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.\n",
      "  return self.fit(X, **fit_params).transform(X)\n",
      "C:\\Users\\zhila.esnaashari\\Anaconda3\\lib\\site-packages\\sklearn\\linear_model\\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.\n",
      "  FutureWarning)\n"
     ]
    }
   ],
   "source": [
    "#########################################OPT4e\n",
    "###Optimal-features Predictor for T4SS Effector proteins\n",
    "###by: Zhila Esna Ashari\n",
    "    \n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "import subprocess\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.preprocessing import scale\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import shutil\n",
    "from shutil import copyfile\n",
    "import tkinter\n",
    "from tkinter import *\n",
    "###################################################add a toggle button to show/hide the code##########################\n",
    "from IPython.display import display\n",
    "from IPython.display import HTML\n",
    "import IPython.core.display as di \n",
    "\n",
    "di.display_html('<script>jQuery(function() {if (jQuery(\"body.notebook_app\").length == 0) { jQuery(\".input_area\").toggle(); jQuery(\".prompt\").toggle();}});</script>', raw=True)\n",
    "\n",
    "di.display_html('''<button onclick=\"jQuery('.input_area').toggle(); jQuery('.prompt').toggle();\">Toggle code</button>''', raw=True)\n",
    "\n",
    "\n",
    "####################################################------------------------------------------############################\n",
    "\n",
    "def calculating_features(filename):\n",
    "    \n",
    "    \n",
    "    # with open('seq3.fasta') as f:\n",
    "    #     lines = f.readlines()\n",
    "    \n",
    "    fasta_file=filename + \".fa\"\n",
    "    lines = [line.rstrip('\\n') for line in open(fasta_file)]\n",
    "\n",
    "    # print(lines)\n",
    "\n",
    "    \n",
    "    my_sequence= ''\n",
    "    sequences= []\n",
    "    sequences_name= []\n",
    "\n",
    "    # my_sequence_name=lines[0]\n",
    "\n",
    "    for l in lines:\n",
    "        if(l == \"\"): #ignore blank lines\n",
    "                pass\n",
    "        elif (l[0] == '>'):\n",
    "            sequences_name.append(l[1:])\n",
    "            sequences.append(my_sequence)\n",
    "            my_sequence= ''\n",
    "        else:\n",
    "            my_sequence+= l\n",
    "\n",
    "    sequences.append(my_sequence)\n",
    "    del sequences[0]\n",
    "\n",
    "    # sequences[1][0]\n",
    "    # print(sequences)\n",
    "    # print(sequences_name[0])\n",
    "    # len(sequences_name)\n",
    "\n",
    "\n",
    "    \n",
    "    # output_features_all=np.zeros((len(sequences_name),1))\n",
    "    output_features_all=pd.DataFrame()\n",
    "    # output_features_all['names']=sequences_name\n",
    "    # print(output_features_all)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    ####################################################------------------------------------------############################\n",
    "    ###FUNCTIONS#######################\n",
    "    \n",
    "\n",
    "    def hydropathy(seq, num):\n",
    "        hydropathy= 0\n",
    "\n",
    "        if (num==1):\n",
    "            seq2=seq\n",
    "        elif(num==2):\n",
    "            seq2=seq[-25:]\n",
    "        elif (num==3):\n",
    "            seq2=seq[0:25]\n",
    "\n",
    "        for amino_acid in seq2:\n",
    "            if (amino_acid=='G'):\n",
    "                hydropathy-=0.4\n",
    "            elif (amino_acid=='A'):\n",
    "                hydropathy+=1.8\n",
    "            elif (amino_acid=='P'):\n",
    "                hydropathy-=1.6\n",
    "            elif (amino_acid=='V'):\n",
    "                hydropathy+=4.2\n",
    "            elif (amino_acid=='L'):\n",
    "                hydropathy+=3.8\n",
    "            elif (amino_acid=='I'):\n",
    "                hydropathy+=4.5\n",
    "            elif (amino_acid=='M'):\n",
    "                hydropathy+=1.9\n",
    "            elif (amino_acid=='F'):\n",
    "                hydropathy+=2.8\n",
    "            elif (amino_acid=='Y'):\n",
    "                hydropathy-=1.3\n",
    "            elif (amino_acid=='W'):\n",
    "                hydropathy-=0.9\n",
    "            elif (amino_acid=='S'):\n",
    "                hydropathy-=0.8\n",
    "            elif (amino_acid=='T'):\n",
    "                hydropathy-=0.7\n",
    "            elif (amino_acid=='C'):\n",
    "                hydropathy+=2.5\n",
    "            elif (amino_acid=='N'):\n",
    "                hydropathy-=3.5\n",
    "            elif (amino_acid=='Q'):\n",
    "                hydropathy-=3.5\n",
    "            elif (amino_acid=='K'):\n",
    "                hydropathy-=3.9\n",
    "            elif (amino_acid=='H'):\n",
    "                hydropathy-=3.2\n",
    "            elif (amino_acid=='R'):\n",
    "                hydropathy-=4.5\n",
    "            elif (amino_acid=='D'):\n",
    "                hydropathy-=3.5\n",
    "            elif (amino_acid=='E'):\n",
    "                hydropathy-=3.5\n",
    "        return hydropathy\n",
    "\n",
    "    ############################\n",
    "    \n",
    "    def Hydropathy_avg(seq, total_hydropathy):\n",
    "        Hydropathy_avg= total_hydropathy/ len(seq)\n",
    "        return Hydropathy_avg\n",
    "\n",
    "    ############################\n",
    "\n",
    "    \n",
    "    def Polarity(seq):\n",
    "        Polarity= 0.0\n",
    "\n",
    "        for amino_acid in seq:\n",
    "            if (amino_acid=='G'):\n",
    "                Polarity+=9\n",
    "            elif (amino_acid=='A'):\n",
    "                Polarity+=8.1\n",
    "            elif (amino_acid=='P'):\n",
    "                Polarity+=8\n",
    "            elif (amino_acid=='V'):\n",
    "                Polarity+=5.9\n",
    "            elif (amino_acid=='L'):\n",
    "                Polarity+=4.9\n",
    "            elif (amino_acid=='I'):\n",
    "                Polarity+=5.2\n",
    "            elif (amino_acid=='M'):\n",
    "                Polarity+=5.7\n",
    "            elif (amino_acid=='F'):\n",
    "                Polarity+=5.2\n",
    "            elif (amino_acid=='Y'):\n",
    "                Polarity+=6.2\n",
    "            elif (amino_acid=='W'):\n",
    "                Polarity+=5.4\n",
    "            elif (amino_acid=='S'):\n",
    "                Polarity+=9.2\n",
    "            elif (amino_acid=='T'):\n",
    "                Polarity+=8.6\n",
    "            elif (amino_acid=='C'):\n",
    "                Polarity+=5.5\n",
    "            elif (amino_acid=='N'):\n",
    "                Polarity+=11.6\n",
    "            elif (amino_acid=='Q'):\n",
    "                Polarity+=10.5\n",
    "            elif (amino_acid=='K'):\n",
    "                Polarity+=11.3\n",
    "            elif (amino_acid=='H'):\n",
    "                Polarity+=10.4\n",
    "            elif (amino_acid=='R'):\n",
    "                Polarity+=10.5\n",
    "            elif (amino_acid=='D'):\n",
    "                Polarity+=13\n",
    "            elif (amino_acid=='E'):\n",
    "                Polarity+=12.3\n",
    "        return Polarity\n",
    "\n",
    "\n",
    "\n",
    "    ############################\n",
    "\n",
    "\n",
    "    \n",
    "    def Mr(seq):\n",
    "        Mr= 0\n",
    "\n",
    "\n",
    "        for amino_acid in seq:\n",
    "            if (amino_acid=='G'):\n",
    "                Mr+=+75.07\n",
    "            elif (amino_acid=='A'):\n",
    "                Mr+=+89.09\n",
    "            elif (amino_acid=='P'):\n",
    "                Mr+=+115.13\n",
    "            elif (amino_acid=='V'):\n",
    "                Mr+=+117.15\n",
    "            elif (amino_acid=='L'):\n",
    "                Mr+=+131.17\n",
    "            elif (amino_acid=='I'):\n",
    "                Mr+=+131.17\n",
    "            elif (amino_acid=='M'):\n",
    "                Mr+=+149.21\n",
    "            elif (amino_acid=='F'):\n",
    "                Mr+=+165.19\n",
    "            elif (amino_acid=='Y'):\n",
    "                Mr+=+181.19\n",
    "            elif (amino_acid=='W'):\n",
    "                Mr+=+204.24\n",
    "            elif (amino_acid=='S'):\n",
    "                Mr+=+105.09\n",
    "            elif (amino_acid=='T'):\n",
    "                Mr+=+119.12\n",
    "            elif (amino_acid=='C'):\n",
    "                Mr+=+121.15\n",
    "            elif (amino_acid=='N'):\n",
    "                Mr+=+132.12\n",
    "            elif (amino_acid=='Q'):\n",
    "                Mr+=+146.15\n",
    "            elif (amino_acid=='K'):\n",
    "                Mr+=+146.19\n",
    "            elif (amino_acid=='H'):\n",
    "                Mr+=+155.16\n",
    "            elif (amino_acid=='R'):\n",
    "                Mr+=+174.20\n",
    "            elif (amino_acid=='D'):\n",
    "                Mr+=+133.10\n",
    "            elif (amino_acid=='E'):\n",
    "                Mr+=+147.13\n",
    "        return Mr\n",
    "\n",
    "\n",
    "    ##########################################################\n",
    "    \n",
    "    def aa_composition(seq,aa_list):\n",
    "        aa_comp= [None] * 20\n",
    "        for i in range(20):\n",
    "            aa_comp[i]=seq.count(aa_list[i])/len(seq)\n",
    "\n",
    "        return aa_comp\n",
    "    #########################################################\n",
    "    \n",
    "    def dp_composition(seq,dp_list):\n",
    "        dp_comp= [None] * 400\n",
    "        for i in range(400):\n",
    "            dp_comp[i]=seq.count(dp_list[i])/(len(seq)-1)\n",
    "        return dp_comp\n",
    "\n",
    "\n",
    "    ############################################################\n",
    "    \n",
    "\n",
    "    def pssm_composition(seq,aa_list,mypssm):\n",
    "        pssm_comp= [0] * 400\n",
    "    #     print(len(pssm_comp))\n",
    "        for i in range(20):\n",
    "            mysum=[0]*20\n",
    "            char_index=[]\n",
    "            for j in range(len(seq)):\n",
    "                if aa_list[i]==seq[j]:\n",
    "                    char_index.append(j)\n",
    "\n",
    "            for k in range(len(char_index)):\n",
    "                mysum=np.add(mysum,list(map(int, mypssm[char_index[k]])))\n",
    "\n",
    "            pssm_comp[20*i:20*(i+1)-1]=mysum\n",
    "    #         print(len(pssm_comp))\n",
    "    #         print(mysum)\n",
    "    #         print(pssm_comp)\n",
    "\n",
    "        pssm_comp=pssm_comp[0:400]\n",
    "        pssm_comp=np.divide(pssm_comp,float(len(seq)))\n",
    "\n",
    "        return pssm_comp\n",
    "\n",
    "\n",
    "    ############################################################\n",
    "    \n",
    "\n",
    "    def pssm_autocov(seq,aa_list,mypssm):\n",
    "        G=8\n",
    "        pssm_auto= [0] * 20 *G  ##########################G=8\n",
    "\n",
    "        for i in range(20):\n",
    "            t=[0]*G\n",
    "            smean=sum(list(map(int, mypssm[:,i])))/len(seq)\n",
    "\n",
    "            for j in range(G):\n",
    "                for k in range(len(seq)-j-1):\n",
    "                    J=(int(mypssm[k,i])-smean)*(int(mypssm[k+j+1,i])-smean)\n",
    "                    t[j]=t[j]+J\n",
    "                t[j]=t[j]/(len(seq)-j)\n",
    "    #         print(t)\n",
    "            pssm_auto[8*i:8*(i+1)-1]=t\n",
    "\n",
    "\n",
    "\n",
    "        pssm_auto=pssm_auto[0:160]\n",
    "        return pssm_auto\n",
    "\n",
    "    ############################################################\n",
    "    ####################################################------------------------------------------############################\n",
    "    #############################Build up output columns############################\n",
    "    for i in range(1,401):\n",
    "        output_features_all[\"V%s\"%i]=\"\"\n",
    "\n",
    "    for i in range(1,161):\n",
    "        output_features_all[\"W%s\"%i]=\"\"\n",
    "\n",
    "    # print(output_features_all)\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    aa_list='ARNDCQEGHILKMFPSTWYV'\n",
    "    for i in aa_list:\n",
    "        output_features_all[\"%s\"%i]=\"\"\n",
    "\n",
    "    dp_list= [None] * 400\n",
    "    n=0\n",
    "    for i in range(20):\n",
    "        for j in range(20):\n",
    "            a=aa_list[i]+aa_list[j]\n",
    "            dp_list[n]=a\n",
    "            n+=1\n",
    "            output_features_all[\"%s\"%a]=\"\"\n",
    "\n",
    "    # print(dp_list)\n",
    "    # print(len(dp_list))\n",
    "\n",
    "    output_features_all[\"total_hydropathy\"]=\"\"\n",
    "    output_features_all[\"hydropathy_C_terminal\"]=\"\"\n",
    "    output_features_all[\"hydropathy_N_terminal\"]=\"\"\n",
    "    output_features_all[\"hydropathy_average\"]=\"\"\n",
    "    output_features_all[\"polarity\"]=\"\"\n",
    "    output_features_all[\"molecular_mass\"]=\"\"\n",
    "    output_features_all[\"homology\"]=\"\"\n",
    "    ##############run functions###################################################################\n",
    "\n",
    "    seqs_length=[]\n",
    "    total_hydropathy=[]\n",
    "    hydropathy_C_terminal=[]\n",
    "    hydropathy_N_terminal=[]\n",
    "    hydropathy_average=[]\n",
    "    polarity=[]\n",
    "    molecular_mass=[]\n",
    "    homology=[]\n",
    "\n",
    "\n",
    "    ##################Run and loop over sequences#########################################################################\n",
    "    counter=0\n",
    "    counting_seq=0\n",
    "\n",
    "    for seq in sequences:\n",
    "\n",
    "    ################################################\n",
    "        seqs_length.append(len(seq))\n",
    "\n",
    "        total_hydropathy.append(hydropathy(seq, 1))\n",
    "        hydropathy_C_terminal.append(hydropathy(seq, 2))\n",
    "        hydropathy_N_terminal.append(hydropathy(seq, 3))\n",
    "        hydropathy_average.append(Hydropathy_avg(seq, total_hydropathy[counter]))\n",
    "        polarity.append(Polarity(seq))\n",
    "        molecular_mass.append(Mr(seq))\n",
    "        counter+=1\n",
    "\n",
    "    ##########################################\n",
    "\n",
    "        aa_comp=aa_composition(seq,aa_list)\n",
    "        dp_comp=dp_composition(seq,dp_list)\n",
    "        output_features_all.at[counting_seq,\"A\":\"V\"]=aa_comp\n",
    "        output_features_all.at[counting_seq,\"AA\":\"VV\"]=dp_comp\n",
    "\n",
    "    #     print(aa_comp)\n",
    "    # #     print(dp_comp)\n",
    "    #     print(seq)\n",
    "    #     print(len(seq))\n",
    "\n",
    "    ###***PSSM***###\n",
    "    ################################################\n",
    "    #####################################################################################################\n",
    "    \n",
    "        x_len=len(seq)\n",
    "        ###run psi-blast to get pssm file\n",
    "        myfile = open(\"seq_pssm.fa\", \"w\")\n",
    "        myfile.write(seq)\n",
    "        myfile.close()\n",
    "        subprocess.call('.\\\\blast\\\\bin\\\\blastpgp -i .\\\\seq_pssm.fa -d uniprot -j 3 -h 0.001 -a 8 -Q .\\\\outname_test.pssm', shell=True)\n",
    "\n",
    "        ####read and parse and extract pssm \n",
    "        xx=open(\"outname_test.pssm\", \"r\")\n",
    "        yy=[line.rstrip('\\n') for line in xx]  \n",
    "\n",
    "        thefile = open('mytemp.txt', 'w')\n",
    "        for item in yy[3:-6]:\n",
    "            thefile.write( item )\n",
    "\n",
    "        # thefile = open('mytemp.txt', 'w')#//////////////////////\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\n",
    "        # for item in yy[3:-6]:\n",
    "        #     thefile.write( item )\n",
    "\n",
    "        thefile.close()#//////////////////\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\n",
    "\n",
    "\n",
    "        input1 = np.loadtxt('mytemp.txt', dtype=(str, float ))\n",
    "    #     print(input1)\n",
    "    #     print(len(input1))\n",
    "        myarray=np.array(input1)\n",
    "        mypssm1=np.resize(myarray,(x_len, 44))\n",
    "        mypssm=mypssm1[:,2:22 ]\n",
    "\n",
    "    #     print(mypssm)\n",
    "#         print(len(mypssm))\n",
    "#         print(input1[-43])\n",
    "        ################################create pssm-comp\n",
    "        pssm_comp=pssm_composition(seq,aa_list,mypssm)\n",
    "    #     print(pssm_comp)\n",
    "    #     print(len(pssm_comp))\n",
    "        ##############################create pssm-auto\n",
    "        pssm_auto=pssm_autocov(seq,aa_list,mypssm)\n",
    "    #     print(pssm_auto)\n",
    "    #     print(len(pssm_auto))\n",
    "\n",
    "    ######################\n",
    "        output_features_all.at[counting_seq,\"V1\":\"V400\"]=pssm_comp\n",
    "        output_features_all.at[counting_seq,\"W1\":\"W160\"]=pssm_auto\n",
    "\n",
    "        counting_seq+=1\n",
    "\n",
    "\n",
    "    ################################################################\n",
    "        myfile = open(\"seq_hom.fa\", \"w\")\n",
    "        myfile.write(seq)\n",
    "        myfile.close()\n",
    "\n",
    "        homol=0\n",
    "        subprocess.call('.\\\\ncbi-blast-2.7.1+\\\\bin\\\\blastp -query .\\\\seq_hom.fa -db effector_db.txt -out .\\\\outname.temp -evalue 0.01', shell=True)\n",
    "\n",
    "        if '>' in open('outname.temp').read():\n",
    "            homol=1\n",
    "#         print (homol)\n",
    "        homology.append(homol)\n",
    "    \n",
    "    ################################################################\n",
    "        subprocess.call(\"pepcoil -sequence .\\\\input_sequences.fa -window 28 -coil -frame -noother -rformat motif -auto -outfile input_emboss.pepcoil\")\n",
    "        ##Coiled Coils\n",
    "        #####Gather coiled coil regions from output file\n",
    "        coiled_coils=[]\n",
    "        with open('input_emboss.pepcoil') as f:###outfile is the output of pepcoil software\n",
    "            lines1 = f.readlines()\n",
    "        ####################################       \n",
    "        for mylines in lines1:\n",
    "            if \"HitCount: \" in mylines:\n",
    "                coiled_coils.append(mylines[12])\n",
    "\n",
    "\n",
    "        #         print(mylines[12])\n",
    "        #         counter+=1\n",
    "        # print(coiled_coils)\n",
    "\n",
    "        # output_features_all[\"Coiled_coils\"]=coiled_coils\n",
    "        # print(output_features_all)\n",
    "\n",
    "    #     pd.DataFrame(coiled_coils).to_csv(\"Coiled_coils_Anap_HGE_merged.csv\", header=False, index=False)\n",
    "\n",
    "        #######################################################\n",
    "        signalp=[]\n",
    "        with open('sig_out') as f2:###outfile is the output of pepcoil software\n",
    "            lines2 = f2.readlines()\n",
    "        ############################################################the specific number we are looking for is in the index 12 of the line       \n",
    "        for mylines in lines2:\n",
    "        #     print(mylines)\n",
    "            if \" N \" in mylines:\n",
    "                signalp.append(0)\n",
    "            elif \" Y \" in mylines:\n",
    "                signalp.append(1)\n",
    "\n",
    "        # # print(signalp)\n",
    "        # pd.DataFrame(signalp).to_csv(\"signalp_train3.csv\", header=False, index=False)\n",
    "        # # output_features_all[\"SignalP\"]=signalp\n",
    "        # # print(output_features_all)\n",
    "\n",
    "\n",
    "    ####################################################------------------------------------------############################\n",
    "    ################################ Edit output (add or remove features)\n",
    "    # print(seqs_length)\n",
    "\n",
    "    output_features_all[\"length\"]=seqs_length\n",
    "    # print(output_features_all)\n",
    "\n",
    "\n",
    "    # print (total_hydropathy)\n",
    "    # print (hydropathy_C_terminal)\n",
    "    # print (hydropathy_N_terminal)\n",
    "    # print (hydropathy_average)\n",
    "    # print (polarity)\n",
    "    # print(molecular_mass)\n",
    "\n",
    "\n",
    "    output_features_all[\"total_hydropathy\"]=total_hydropathy\n",
    "    output_features_all[\"hydropathy_C_terminal\"]=hydropathy_C_terminal\n",
    "    output_features_all[\"hydropathy_N_terminal\"]=hydropathy_N_terminal\n",
    "    output_features_all[\"hydropathy_average\"]=hydropathy_average\n",
    "    output_features_all[\"polarity\"]=polarity\n",
    "    output_features_all[\"molecular_mass\"]=molecular_mass\n",
    "    output_features_all[\"homology\"]=homology\n",
    "    output_features_all[\"coiled_coil\"]=coiled_coils\n",
    "    # print(output_features_all)\n",
    "\n",
    "\n",
    "\n",
    "    ########################\n",
    "    #######\n",
    "\n",
    "    ###create a list of them to remove (this is for dp)\n",
    "    removing= []\n",
    "    selected=[\"KE\",\"EK\",\"AG\",\"KK\",\"GG\",\"VG\",\"EE\",\"KD\",\"DK\",\"KL\",\"GL\",\"VI\",\"VV\",\"GA\",\"NN\",\"GV\",\"FF\",\"LP\",\"AV\",\"VR\",\"KI\",\"AR\",\"NS\",\"IA\",\"LA\",\"WQ\",\"PL\",\"LE\",\"RA\",\"AW\",\"EL\",\"FS\",\"LG\",\"DS\",\"SS\",\"QV\",\"SD\",\"NL\",\"TN\",\"GS\",\"IK\",\"VW\",\"AA\",\"MA\",\"MG\",\"YL\"]\n",
    "\n",
    "    n=0\n",
    "    for i in range(20):\n",
    "        for j in range(20):\n",
    "            a=aa_list[i]+aa_list[j]\n",
    "            if a not in selected:\n",
    "                removing.append(a)\n",
    "            n+=1\n",
    "\n",
    "    ######\n",
    "\n",
    "    output_features_all=output_features_all.drop(['C', 'Q','H','I','L','P','T','W','Y'], axis=1) ###remove for aa\n",
    "    output_features_all=output_features_all.drop(removing, axis=1)###remove for dp\n",
    "\n",
    "    # print(output_features_all.shape)   \n",
    "    # print(output_features_all)\n",
    "\n",
    "    ########################\n",
    "\n",
    "    ########################\n",
    "    ###\n",
    "    removing_pssm_comp= []\n",
    "    removing_pssm_auto= []\n",
    "    selected_pssm_comp=[\"V5\",\"V303\",\"V246\",\"V113\",\"V143\",\"V49\",\"V56\",\"V15\",\"V205\",\"V397\",\"V20\",\"V232\",\"V158\",\"V323\",\"V17\",\"V23\",\"V260\",\"V156\",\"V369\",\"V308\",\"V227\",\"V278\",\"V132\",\"V399\",\"V43\",\"V144\",\"V126\",\"V136\",\"V312\",\"V41\",\"V179\",\"V2\",\"V340\",\"V222\",\"V8\",\"V29\",\"V385\",\"V12\",\"V226\",\"V236\",\"V306\",\"V251\",\"V76\",\"V145\",\"V230\",\"V304\",\"V221\",\"V10\",\"V7\",\"V33\",\"V6\",\"V155\",\"V395\",\"V124\",\"V381\",\"V237\",\"V386\",\"V228\",\"V13\",\"V279\",\"V16\",\"V392\",\"V127\",\"V149\",\"V229\",\"V224\",\"V378\",\"V393\",\"V307\",\"V387\",\"V14\",\"V154\",\"V253\",\"V18\",\"V11\",\"V57\",\"V159\",\"V142\",\"V42\",\"V280\",\"V384\",\"V234\",\"V48\",\"V19\",\"V147\",\"V4\",\"V9\",\"V240\",\"V121\",\"V67\",\"V231\",\"V26\",\"V160\",\"V286\",\"V239\",\"V27\",\"V388\",\"V250\",\"V274\",\"V60\",\"V146\",\"V225\",\"V271\",\"V379\",\"V141\",\"V153\",\"V223\",\"V383\",\"V64\",\"V238\",\"V374\",\"V107\",\"V44\",\"V63\",\"V389\",\"V66\",\"V51\",\"V58\",\"V125\",\"V152\",\"V157\",\"V398\",\"V31\",\"V37\",\"V62\",\"V150\",\"V61\",\"V50\",\"V382\",\"V297\",\"V35\",\"V131\",\"V151\",\"V130\",\"V194\",\"V3\",\"V55\",\"V248\",\"V28\",\"V396\",\"V254\",\"V321\",\"V128\",\"V54\",\"V38\",\"V25\",\"V275\",\"V137\",\"V52\",\"V112\",\"V261\",\"V242\",\"V72\",\"V241\",\"V285\",\"V68\",\"V269\",\"V252\",\"V284\",\"V21\",\"V348\",\"V293\",\"V258\",\"V39\",\"V287\",\"V266\",\"V244\",\"V292\",\"V346\",\"V361\",\"V45\",\"V140\",\"V262\",\"V34\",\"V134\",\"V288\",\"V277\",\"V371\",\"V366\",\"V394\",\"V289\",\"V71\",\"V53\",\"V133\",\"V245\",\"V59\",\"V300\",\"V257\",\"V249\",\"V214\",\"V377\",\"V36\",\"V75\",\"V391\",\"V350\",\"V148\",\"V129\",\"V122\",\"V317\",\"V301\",\"V316\",\"V69\",\"V165\",\"V400\",\"V178\",\"V256\",\"V175\",\"V65\",\"V174\",\"V77\",\"V70\",\"V243\",\"V168\",\"V259\",\"V167\",\"V247\",\"V74\",\"V79\",\"V73\",\"V233\",\"V80\",\"V78\",\"V200\",\"V185\",\"V196\",\"V195\",\"V183\",\"V184\",\"V192\",\"V188\",\"V182\",\"V198\",\"V181\"]\n",
    "    selected_pssm_auto=[\"W12\",\"W92\",\"W113\",\"W11\",\"W91\",\"W51\",\"W44\",\"W43\",\"W52\",\"W49\",\"W25\",\"W89\",\"W41\",\"W27\",\"W68\",\"W115\",\"W10\",\"W114\",\"W9\",\"W140\",\"W66\",\"W90\",\"W57\",\"W35\",\"W67\",\"W137\",\"W26\",\"W142\",\"W42\",\"W139\",\"W18\",\"W4\",\"W65\",\"W138\",\"W28\",\"W34\",\"W116\",\"W19\",\"W50\",\"W17\",\"W1\",\"W122\",\"W20\",\"W148\",\"W33\",\"W13\",\"W124\",\"W130\",\"W121\",\"W93\",\"W36\",\"W141\",\"W58\",\"W108\",\"W88\",\"W53\",\"W45\",\"W145\",\"W132\",\"W32\",\"W37\",\"W157\",\"W129\",\"W24\",\"W104\"]\n",
    "    n=0\n",
    "    for i in range(1,401):\n",
    "        a=\"V%s\"%i\n",
    "        if a not in selected_pssm_comp:\n",
    "            removing_pssm_comp.append(a)\n",
    "    # print(removing_pssm_comp)\n",
    "\n",
    "    for j in range(1,161):\n",
    "        a=\"W%s\"%j\n",
    "        if a not in selected_pssm_auto:\n",
    "            removing_pssm_auto.append(a)\n",
    "    # print(removing_pssm_auto)\n",
    "    ######\n",
    "\n",
    "    output_features_all=output_features_all.drop(removing_pssm_comp, axis=1)###remove for pssm_comp\n",
    "    output_features_all=output_features_all.drop(removing_pssm_auto, axis=1)###remove for pssm_auto\n",
    "\n",
    "#     print(output_features_all.shape)   \n",
    "#     print(output_features_all)\n",
    "    output_features_all.to_csv(\"%s.csv\" %filename, header=True, index=False)\n",
    "    \n",
    "    fasta_file3=\"input_sequences.fa\"\n",
    "    lines3 = [line.rstrip('\\n') for line in open(fasta_file3)]\n",
    "\n",
    "    my_sequence3= ''\n",
    "    sequences3= []\n",
    "    sequences_name3= []\n",
    "\n",
    "    for l in lines3:\n",
    "            if(l == \"\"): #ignore blank lines\n",
    "                    pass\n",
    "            elif (l[0] == '>'):\n",
    "                sequences_name3.append(l[0:])\n",
    "                sequences3.append(my_sequence3)\n",
    "                my_sequence3= ''\n",
    "            else:\n",
    "                my_sequence3+= l\n",
    "\n",
    "    sequences3.append(my_sequence3)\n",
    "    del sequences3[0]\n",
    "    \n",
    "    features_print=pd.read_csv('input_sequences.csv')\n",
    "\n",
    "    features_print2=features_print[['hydropathy_C_terminal','hydropathy_average' ,'homology', 'length', 'coiled_coil']]\n",
    "\n",
    "    features_print2['name']=sequences_name3\n",
    "    pd.DataFrame(features_print2).to_csv(\"SOME_FEATURES_OF_INPUT_SEQS.csv\", index=False)\n",
    "\n",
    "    \n",
    "    \n",
    "#########################################     Prediction     ######################################  \n",
    "\n",
    "def predict_candidate_effectors():\n",
    "    \n",
    "    ###READ trainig and test set and extract lables\n",
    "    training_set1=pd.read_csv('training_set.csv')\n",
    "    training_set=training_set1.iloc[:, 0:364]\n",
    "    label=training_set1.iloc[:,364]\n",
    "    training_rows=training_set.shape[0]\n",
    "    \n",
    "    testing_set=pd.read_csv('input_sequences.csv')\n",
    "    testing_rows1=testing_set.shape[0]\n",
    "    \n",
    "    if testing_rows1<21:\n",
    "        for i in range(training_rows):\n",
    "            testing_set=testing_set.append(training_set.iloc[i,:])\n",
    "        testing_set=testing_set.reset_index(drop=True)\n",
    "\n",
    "\n",
    "    ##NORMALIZE\n",
    "    scaler = StandardScaler()\n",
    "    training_set = pd.DataFrame(scaler.fit_transform(training_set), index=training_set.index, columns=training_set.columns)\n",
    "    testing_set = pd.DataFrame(scaler.transform(testing_set), index=testing_set.index, columns=testing_set.columns)\n",
    "    \n",
    "    \n",
    "    if testing_rows1<21:\n",
    "        testing_rows=testing_set.shape[0]\n",
    "        for i in range(testing_rows1,testing_rows):\n",
    "            testing_set=testing_set.drop(i)\n",
    "\n",
    "    ###Predict effectors using machine learning\n",
    "\n",
    "    svm_model_linear = SVC(kernel = 'rbf', C = 1).fit(training_set, label)\n",
    "    predictions = svm_model_linear.predict(testing_set)\n",
    "\n",
    "    svm_model_linear = SVC(kernel = 'linear', C = 1).fit(training_set, label)\n",
    "    predictions2 = svm_model_linear.predict(testing_set)\n",
    "\n",
    "    logistic_model=LogisticRegression().fit(training_set, label)\n",
    "    predictions3 = logistic_model.predict(testing_set)\n",
    "\n",
    "#     pd.DataFrame(predictions).to_csv(\"predict_svmrbf.csv\", header=True, index=False)\n",
    "#     pd.DataFrame(predictions2).to_csv(\"predict_svmlinear.csv\", header=True, index=False)\n",
    "#     pd.DataFrame(predictions3).to_csv(\"predict_LR.csv\", header=True, index=False)\n",
    "\n",
    "#########################################################################################\n",
    "#########################################################################################\n",
    "#########################################################################################\n",
    "#################################################################\n",
    "    lines = [line.rstrip('\\n') for line in open('training_set_non-effectors.fasta')]\n",
    "    # lines = [line.rstrip('\\n') for line in open('Anaplasma_Phagocytophilum_effectors.fasta')]\n",
    "    # print(lines)\n",
    "\n",
    "    ###read all sequences ##############################################\n",
    "    my_sequence= ''\n",
    "    sequences_noneff= []\n",
    "    sequences_name_noneff= []\n",
    "\n",
    "\n",
    "    for l in lines:\n",
    "        if(l == \"\"): #ignore blank lines\n",
    "                pass\n",
    "        elif (l[0] == '>'):\n",
    "            sequences_name_noneff.append(l[1:])\n",
    "            sequences_noneff.append(my_sequence)\n",
    "            my_sequence= ''\n",
    "        else:\n",
    "            my_sequence+= l\n",
    "\n",
    "    sequences_noneff.append(my_sequence)\n",
    "    del sequences_noneff[0]\n",
    "\n",
    "\n",
    "# #################################################################\n",
    "    lines = [line.rstrip('\\n') for line in open('input_sequences.fa')]\n",
    "    # lines = [line.rstrip('\\n') for line in open('Anaplasma_Phagocytophilum_effectors.fasta')]\n",
    "    # print(lines)\n",
    "\n",
    "    ###read all sequences in sequences and their names in sequences_name##############################################\n",
    "    my_sequence= ''\n",
    "    sequences_input= []\n",
    "    sequences_name_input= []\n",
    "\n",
    "\n",
    "    for l in lines:\n",
    "        if(l == \"\"): #ignore blank lines\n",
    "                pass\n",
    "        elif (l[0] == '>'):\n",
    "            sequences_name_input.append(l[1:])\n",
    "            sequences_input.append(my_sequence)\n",
    "            my_sequence= ''\n",
    "        else:\n",
    "            my_sequence+= l\n",
    "\n",
    "    sequences_input.append(my_sequence)\n",
    "    del sequences_input[0]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#########################################################################################\n",
    "#########################################################################################\n",
    "#########################################################################################\n",
    "##################################################### (predictions analysis)#################################\n",
    "    predict_file = open(\"predicted_effectors_total.fasta\", \"w+\")\n",
    "\n",
    "    count_effector=0   \n",
    "    for i in range(len(predictions)):\n",
    "        if (predictions[i]==1 and sequences_input[i] not in sequences_noneff):\n",
    "            predict_file.write(str(sequences_name_input[i])+ '\\n')\n",
    "            predict_file.write(str(sequences_input[i])+ '\\n'+ '\\n')\n",
    "            count_effector+=1\n",
    "    predict_file.close()\n",
    "    del predict_file\n",
    "#     print(count_effector)\n",
    "\n",
    "    one_line=\"number of predicted effectors(total)=\"+str(count_effector) + \"\\n\"  +\"\\n\" \n",
    "    with open(\"predicted_effectors_total.fasta\", 'r+') as fp:\n",
    "        lines = fp.readlines()     # lines is list of line, each element '...\\n'\n",
    "        lines.insert(0, one_line)  # you can use any index if you know the line index\n",
    "        fp.seek(0)                 # file pointer locates at the beginning to write the whole file again\n",
    "        fp.writelines(lines)  \n",
    "\n",
    "\n",
    "    #########################  see which one has most probability based on voting\n",
    "    predict_file1 = open(\"predicted_effectors_mostlikely.fasta\", \"w+\")\n",
    "    predict_file2 = open(\"predicted_effectors_probable.fasta\", \"w+\")\n",
    "    predict_file3 = open(\"predicted_effectors_lesslikely.fasta\", \"w+\")\n",
    "\n",
    "    count_effector_most=0\n",
    "    for i in range(len(predictions)):\n",
    "        if (predictions[i]==1 and sequences_input[i] not in sequences_noneff):\n",
    "            if (predictions2[i] ==1 and predictions3[i]==1):\n",
    "                predict_file1.write(str(sequences_name_input[i])+ '\\n')\n",
    "                predict_file1.write(str(sequences_input[i])+ '\\n'+ '\\n')\n",
    "                count_effector_most+=1\n",
    "\n",
    "\n",
    "    predict_file1.close()\n",
    "    del predict_file1\n",
    "#     print(\"most:\")\n",
    "#     print(count_effector_most)\n",
    "\n",
    "    one_line=\"number of predicted effectors(most likely)=\"+str(count_effector_most) + \"\\n\"  +\"\\n\" \n",
    "    with open(\"predicted_effectors_mostlikely.fasta\", 'r+') as fp:\n",
    "        lines = fp.readlines()     # lines is list of line, each element '...\\n'\n",
    "        lines.insert(0, one_line)  # you can use any index if you know the line index\n",
    "        fp.seek(0)                 # file pointer locates at the beginning to write the whole file again\n",
    "        fp.writelines(lines) \n",
    "    ###\n",
    "\n",
    "    count_effector_mid=0\n",
    "    for i in range(len(predictions)):\n",
    "        if (predictions[i]==1 and sequences_input[i] not in sequences_noneff):\n",
    "            if ((predictions2[i] ==1 and predictions3[i]==-1) or (predictions2[i] ==-1 and predictions3[i]==1)):\n",
    "                predict_file2.write(str(sequences_name_input[i])+ '\\n')\n",
    "                predict_file2.write(str(sequences_input[i])+ '\\n'+ '\\n')\n",
    "                count_effector_mid+=1\n",
    "\n",
    "    predict_file2.close()\n",
    "    del predict_file2\n",
    "#     print(\"mid:\")\n",
    "#     print(count_effector_mid)\n",
    "    one_line=\"number of predicted effectors(probable)=\"+str(count_effector_mid) + \"\\n\"  +\"\\n\" \n",
    "    with open(\"predicted_effectors_probable.fasta\", 'r+') as fp:\n",
    "        lines = fp.readlines()     # lines is list of line, each element '...\\n'\n",
    "        lines.insert(0, one_line)  # you can use any index if you know the line index\n",
    "        fp.seek(0)                 # file pointer locates at the beginning to write the whole file again\n",
    "        fp.writelines(lines) \n",
    "    ###\n",
    "\n",
    "    count_effector_less=0\n",
    "    for i in range(len(predictions)):\n",
    "        if (predictions[i]==1 and sequences_input[i] not in sequences_noneff):\n",
    "            if (predictions2[i] ==-1 and predictions3[i]==-1):\n",
    "                predict_file3.write(str(sequences_name_input[i])+ '\\n')\n",
    "                predict_file3.write(str(sequences_input[i])+ '\\n'+ '\\n')\n",
    "                count_effector_less+=1\n",
    "\n",
    "    predict_file3.close()\n",
    "    del predict_file3\n",
    "#     print(\"less:\")\n",
    "#     print(count_effector_less)  \n",
    "    one_line=\"number of predicted effectors(less likely)=\"+str(count_effector_less) + \"\\n\"  +\"\\n\" \n",
    "    with open(\"predicted_effectors_lesslikely.fasta\", 'r+') as fp:\n",
    "        lines = fp.readlines()     # lines is list of line, each element '...\\n'\n",
    "        lines.insert(0, one_line)  # you can use any index if you know the line index\n",
    "        fp.seek(0)                 # file pointer locates at the beginning to write the whole file again\n",
    "        fp.writelines(lines) \n",
    "\n",
    "\n",
    "#########################################     GUI     ######################################\n",
    "\n",
    "top = tkinter.Tk()\n",
    "top.title(\"OPT4e\")\n",
    "top.geometry(\"500x470\")\n",
    "top.configure(background=\"blue\")\n",
    "\n",
    "\n",
    "canvas = Canvas(top, width=235, height = 139)      \n",
    "canvas.pack(padx=5, pady=5)      \n",
    "# # photo1=PhotoImage(file=)\n",
    "img = PhotoImage(file=\"logo.gif\")      \n",
    "canvas.create_image(120,71, anchor=CENTER, image=img) \n",
    "\n",
    "##############################\n",
    "\n",
    "# # Label(top, image=photo1, bg=\"black\").grid(row=0, column=0, sticky=w)\n",
    "# thelable=Label(top, text=\"OPT4e\", bg=\"royalblue\", fg=\"white\", font=\"none 20 bold\")\n",
    "# # .grid(row=0,column=0)\n",
    "# thelable.pack()\n",
    "# Code to add widgets will go here...\n",
    "\n",
    "# window = Toplevel(top)###open a new window when a button is pressed\n",
    "\n",
    "# # root = Tk()\n",
    "T = Text(top,bg=\"dark blue\", fg=\"white\", font=\"none 12 bold\",bd=7,height=2, width=33, padx=10, pady=10)\n",
    "# insert_centered(T, 'Thinking...')\n",
    "\n",
    "\n",
    "T.pack(padx=10, pady=10)\n",
    "T.insert(END, \"Optimal-features Predictor for T4SS                         Effector proteins   \",\"center\")\n",
    "# # # mainloop()\n",
    "\n",
    "\n",
    "def create_window1():\n",
    "    window = Toplevel(top)\n",
    "    window.geometry(\"320x60\")\n",
    "    window.configure(background=\"grey\")\n",
    "    L1 = Label(window, bg=\"grey\",fg=\"white\",text=\"Results stored in predicted_effectors... files \\nThank you\")\n",
    "    L1.pack( )\n",
    "    \n",
    "    B1=Button(window, text=\"Okay\", command=window.destroy) \n",
    "    B1.pack()\n",
    "    \n",
    "    \n",
    "def create_window2():\n",
    "    window = Toplevel(top)\n",
    "    window.geometry(\"320x60\")\n",
    "    window.configure(background=\"grey\")\n",
    "    L1 = Label(window, bg=\"grey\",fg=\"white\",text=\"New effector sequences added to software \\nThank you\")\n",
    "    L1.pack( )\n",
    "    \n",
    "    B1=Button(window, text=\"Okay\", command=window.destroy) \n",
    "    B1.pack()\n",
    "    \n",
    "def create_window3():\n",
    "    window = Toplevel(top)\n",
    "    window.geometry(\"320x60\")\n",
    "    window.configure(background=\"grey\")\n",
    "    L1 = Label(window, bg=\"grey\",fg=\"white\",text=\"New non-effector sequences added to software \\nThank you\")\n",
    "    L1.pack( )\n",
    "    \n",
    "    B1=Button(window, text=\"Okay\", command=window.destroy) \n",
    "    B1.pack()\n",
    "    \n",
    "def create_window4():\n",
    "    window = Toplevel(top)\n",
    "    window.geometry(\"320x60\")\n",
    "    window.configure(background=\"grey\")\n",
    "    L1 = Label(window, bg=\"grey\",fg=\"white\",text=\"Training set restored to original version. \\nThank you\")\n",
    "    L1.pack( )\n",
    "    \n",
    "    B1=Button(window, text=\"Okay\", command=window.destroy) \n",
    "    B1.pack()\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "def first_button():\n",
    "    calculating_features('input_sequences')\n",
    "    predict_candidate_effectors()\n",
    "    create_window1()\n",
    "  \n",
    "    \n",
    "def second_button():\n",
    "    \n",
    "    file_new = open(\"input_sequences.fa\", \"r\")\n",
    "    data_new = file_new.read()\n",
    "    file_new.close()\n",
    "\n",
    "    file_all= open(\"training_set_effectors.fasta\",\"a\")\n",
    "    file_all.write(\"\\n\")\n",
    "    file_all.write(data_new)\n",
    "    file_all.write(\"\\n\")\n",
    "    file_all.close()\n",
    "    \n",
    "    file_all_db= open(\"effector_db.txt\",\"a\")\n",
    "    file_all_db.write(\"\\n\")\n",
    "    file_all_db.write(data_new)\n",
    "    file_all_db.write(\"\\n\")\n",
    "    file_all_db.close()\n",
    "    \n",
    "    calculating_features('input_sequences')\n",
    "    \n",
    "    \n",
    "    df1 = pd.read_csv('input_sequences.csv')\n",
    "    l = [1] * df1.shape[0]\n",
    "    df1[\"label\"]=l\n",
    "    df1.to_csv(\"seq_NoHeader.csv\", header=False, index=False)\n",
    "    \n",
    "    file_new_features = open(\"seq_NoHeader.csv\", \"r\")\n",
    "    data_new_features = file_new_features.read()\n",
    "    file_new_features.close()\n",
    "    \n",
    "    file_all_features= open(\"training_set.csv\",\"a\")\n",
    "    file_all_features.write(data_new_features)\n",
    "    file_all_features.close()\n",
    "    ###ONE TIME####\n",
    "    ###build up database for BLAST and homology\n",
    "    subprocess.call('.\\\\ncbi-blast-2.7.1+\\\\bin\\\\makeblastdb -in effector_db.txt -dbtype prot', shell=True)\n",
    "    create_window2() \n",
    "    \n",
    "    \n",
    "    \n",
    "def third_button():\n",
    "    file_new = open(\"input_sequences.fa\", \"r\")\n",
    "    data_new = file_new.read()\n",
    "    file_new.close()\n",
    "    \n",
    "    file_all= open(\"training_set_non-effectors.fasta\",\"a\")\n",
    "    file_all.write(\"\\n\")\n",
    "    file_all.write(data_new)\n",
    "    file_all.write(\"\\n\")\n",
    "    file_all.close()\n",
    "    \n",
    "    calculating_features('seq')\n",
    "    \n",
    "    \n",
    "    df1 = pd.read_csv('seq.csv')\n",
    "    l = [-1] * df1.shape[0]\n",
    "    df1[\"label\"]=l\n",
    "    df1.to_csv(\"seq_NoHeader.csv\", header=False, index=False)\n",
    "    \n",
    "    file_new_features = open(\"seq_NoHeader.csv\", \"r\")\n",
    "    data_new_features = file_new_features.read()\n",
    "    file_new_features.close()\n",
    "    \n",
    "    file_all_features= open(\"training_set.csv\",\"a\")\n",
    "    file_all_features.write(data_new_features)\n",
    "    file_all_features.close()\n",
    "        \n",
    "    create_window2() \n",
    "    \n",
    "    \n",
    "def fourth_button():\n",
    "#     os.remove(\"seq3.csv\")\n",
    "#     os.remove(\"seq3.fasta\")\n",
    "    copyfile(\"training_set_effectors_original.fasta\", \"training_set_effectors.fasta\")\n",
    "    copyfile(\"training_set_non-effectors_original.fasta\", \"training_set_non-effectors.fasta\")\n",
    "    copyfile(\"training_set_original.csv\", \"training_set.csv\")\n",
    "    copyfile(\"effector_db_original.txt\", \"effector_db.txt\")\n",
    "    subprocess.call('.\\\\ncbi-blast-2.7.1+\\\\bin\\\\makeblastdb -in effector_db.txt -dbtype prot', shell=True)\n",
    "    create_window4()  \n",
    "    \n",
    "# b = Button(text=\"Create new window\", command=create_window)\n",
    "# b.pack()\n",
    "\n",
    "w = tkinter.Button ( text =\"Predict effector candidates \\n(provide sequences in input_sequences.fa)\", activebackground=\"grey\",command=first_button )\n",
    "# .grid(row=1, column=1, padx=10, pady=10)\n",
    "# w .grid(row=0, column=1, padx=10, pady=10)\n",
    "w.pack(padx=10, pady=10)\n",
    "\n",
    "w2 = tkinter.Button ( text =\"Add new validated effectors to software \\n(provide sequences in input_sequences.fa)\" , activebackground=\"grey\",command=second_button)\n",
    "# .grid(row=2, column=1, padx=10, pady=10)\n",
    "w2.pack(padx=10, pady=10)\n",
    "\n",
    "w3 = tkinter.Button ( text =\"Add new known non-effectors to software \\n(provide sequences in input_sequences.fa)\" , activebackground=\"grey\",command=third_button)\n",
    "# .grid(row=2, column=1, padx=10, pady=10)\n",
    "w3.pack(padx=10, pady=10)\n",
    "\n",
    "w4 = tkinter.Button ( text =\"Reset software to original version\", activebackground=\"grey\",command=fourth_button)\n",
    "w4.pack(padx=10, pady=10)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "top.mainloop()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: '55063.0000'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-5-a2e67d6d5626>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'55063.0000'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m: invalid literal for int() with base 10: '55063.0000'"
     ]
    }
   ],
   "source": [
    "int(('55063.0000'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
